# <center>Đồ án 3: Linear Regression</center>

# Thông tin sinh viên

- Họ và tên: 
- MSSV: 
- Lớp: 

# Import

In [77]:
import pandas as pd
import numpy as np

# Import thêm dữ thư viện nếu cần
from sklearn.model_selection import KFold

# Đọc dữ liệu

In [78]:
# Đọc dữ liệu bằng pandas
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

# Lấy các đặc trưng X và giá trị mục tiêu y cho các tập huấn luyện (train) và kiểm tra (test)
X_train = train.iloc[:, :-1]    # Dataframe (chứa các đặc trưng huấn luyện)
y_train = train.iloc[:, -1]     # Series    (chứa 1 giá trị mục tiêu kiểm tra)

X_test = test.iloc[:, :-1]      # Dataframe (chứa các đặc trưng kiểm tra)
y_test = test.iloc[:, -1]       # Series    (chứa 1 giá trị mục tiêu kiểm tra)

# print('X\n', X_test)
# print('y\n', y_test)

# Sinh viên có thể sử dụng các khác nếu cần

# Cài đặt hàm

In [79]:
# Cài đặt các hàm cần thiết ở đây
class OLSLinearRegression:
    def fit(self, X, y):
        X_pinv = np.linalg.inv(X.T @ X) @ X.T    # np.linalg.pinv(X)
        self.w = X_pinv @ y

        return self

    def get_params(self):
        return self.w

    def predict(self, X):
        return np.sum(self.w.ravel() * X, axis=1) 

def model_mae(y_test, y_pred):
    if len(y_test) != len(y_pred):
        raise ValueError("Input arrays must have the same length.")
    
    absolute_errors = [abs(actual - predicted) for actual, predicted in zip(y_test, y_pred)]
    mae = sum(absolute_errors) / len(y_test)
    return mae

In [80]:
print('Input rain data:')
train.head()

Input rain data:


Gender  10percentage  12percentage  CollegeTier  Degree  collegeGPA  \
0       2         87.80         84.00            1       1       73.82   
1       1         57.00         64.50            2       1       65.00   
2       1         77.33         85.17            2       1       61.94   
3       1         84.30         86.00            1       1       80.40   
4       2         82.00         75.00            2       1       64.30   

   CollegeCityTier  English  Logical  Quant  ...  MechanicalEngg  \
0                1      650      665    810  ...              -1   
1                0      440      435    210  ...              -1   
2                0      485      475    505  ...              -1   
3                1      675      620    635  ...              -1   
4                1      575      495    365  ...              -1   

   ElectricalEngg  TelecomEngg  CivilEngg  conscientiousness  agreeableness  \
0              -1           -1         -1            -0.1590         0.3789   
1              -1           -1         -1             1.1336         0.0459   
2              -1          260         -1             0.5100        -0.1232   
3              -1           -1         -1            -0.4463         0.2124   
4              -1           -1         -1            -1.4992        -0.7473   

   extraversion  nueroticism  openess_to_experience  Salary  
0        1.2396      0.14590                 0.2889  445000  
1        1.2396      0.52620                -0.2859  110000  
2        1.5428     -0.29020                -0.2875  255000  
3        0.3174      0.27270                 0.4805  420000  
4       -1.0697      0.06223                 0.1864  200000  

[5 rows x 24 columns]

In [81]:
print('Input test data:')
test.head()

Input test data:


Gender  10percentage  12percentage  CollegeTier  Degree  collegeGPA  \
0       1         61.63         53.17            2       1       60.00   
1       1         72.40         73.40            2       1       68.25   
2       1         79.00         83.40            2       1       73.30   
3       1         73.00         69.00            2       1       65.50   
4       1         92.00         72.00            2       1       66.00   

   CollegeCityTier  English  Logical  Quant  ...  MechanicalEngg  \
0                1      485      415    215  ...              -1   
1                0      485      625    655  ...              -1   
2                0      425      515    595  ...              -1   
3                0      465      480    415  ...              -1   
4                1      500      440    500  ...              -1   

   ElectricalEngg  TelecomEngg  CivilEngg  conscientiousness  agreeableness  \
0              -1           -1         -1            -0.7264         0.1888   
1              -1           -1         -1            -0.2628        -0.1232   
2              -1           -1         -1             1.4374         0.9688   
3              -1           -1         -1             0.1282        -0.1206   
4              -1           -1         -1            -2.4570        -0.9531   

   extraversion  nueroticism  openess_to_experience  Salary  
0       -0.7794      -0.7603                 0.3049  280000  
1        0.8171       0.5323                 0.0284  520000  
2        0.8171       0.4442                 0.9763  150000  
3        0.3174       0.6530                -0.6692  180000  
4       -0.7585       0.9066                 0.0973  300000  

[5 rows x 24 columns]

# Yêu cầu 1a: Sử dụng toàn bộ 11 đặc trưng đầu tiên `Gender`, `10percentage`, `12percentage`, `CollegeTier`, `Degree`, `collegeGPA`, `CollegeCityTier`, `English`, `Logical`, `Quant`, `Domain` (2 điểm) 

In [82]:
# Phần code cho yêu cầu 1a
X_train1a = X_train[[
    'Gender',
    '10percentage',
    '12percentage',
    'CollegeTier',
    'Degree', 
    'collegeGPA',
    'CollegeCityTier',
    'English',
    'Logical',
    'Quant',
    'Domain',
]]

X_test1a = X_test[[
    'Gender', 
    '10percentage', 
    '12percentage', 
    'CollegeTier', 
    'Degree', 
    'collegeGPA',
    'CollegeCityTier',
    'English',
    'Logical',
    'Quant',
    'Domain',
]]

# To train 11 fearures
def train_1a(X_train, y_train, X_test):
    w = OLSLinearRegression().fit(X_train, y_train)
    y_pred = w.predict(X_test)
    return y_pred, w

def question_1a_MAE_check(y_test, y_pred, w):
    return model_mae(y_test, y_pred), np.array(w.get_params()).reshape(-1, 1)

In [83]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra
y_pred, w = train_1a(X_train1a.to_numpy(), y_train.to_numpy(), X_test1a.to_numpy())
mae, w = question_1a_MAE_check(y_test.to_numpy(), y_pred, w)

print("Weight:\n", np.round(w, 3), "\nMAE:", np.round(mae, 3))

Weight:
 [[-22756.513]
 [   804.503]
 [  1294.655]
 [-91781.898]
 [ 23182.389]
 [  1437.549]
 [ -8570.662]
 [   147.858]
 [   152.888]
 [   117.222]
 [ 34552.286]] 
MAE: 104863.778


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = (-22756.513) * Gender + 804.503 * 10percentage + 1294.655 * 12percentage \\
+ (-91781.898) * CollegeTier + 23182.389 * Degree + 1437.549 * collegeGPA \\
+ (-8570.662) * CollegeCityTier + 147.858 * English + 152.888 * Logical 
\\+ 117.222 * Quant + 34552.286 * Domain $$

# Yêu cầu 1b: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng tính cách với các đặc trưng tính cách gồm `conscientiousness`, `agreeableness`, `extraversion`, `nueroticism`, `openess_to_experience`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [84]:
# Phần code cho yêu cầu 1b
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu

features_1b = [
    'conscientiousness',
    'agreeableness',
    'extraversion',
    'nueroticism',
    'openess_to_experience',
    'Salary'
]

def question_1b_handler(train, features, k_fold=5):
    mae_arr = np.zeros(train.shape[1] - 1)
    shuffle_data = KFold(n_splits = k_fold, shuffle = True)

    for train_split, test_split in shuffle_data.split(train):
        i = 0
        for column in features:
            if column != 'Salary':
                feature_train = train.iloc[train_split].loc[:,[column]]
                label_train = train.iloc[train_split].loc[:,['Salary']]

                feature_test = train.iloc[test_split].loc[:,[column]]
                label_test = train.iloc[test_split].loc[:,['Salary']]
    
                y_pred, w = train_1a(feature_train.to_numpy(), label_train.to_numpy(), feature_test.to_numpy())
                mae, w = question_1a_MAE_check(label_test.to_numpy(), y_pred, w)

                mae_arr[i] += mae
                i += 1

    mae_arr /= k_fold
    print('MAE: ', np.round(mae_arr, 3))

    return np.round(mae_arr, 3), features[np.argmin(mae_arr)]

In [85]:
train_1b = train[features_1b]
mae_arr_1b, best_feature_1b = question_1b_handler(train_1b, features_1b, 5)

print("Best feature: ", best_feature_1b)

MAE:  [306176.936 300721.863 306838.547 299382.744 302932.457]
Best feature:  nueroticism


In [86]:
# Huấn luyện lại mô hình best_personality_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện
features_train_1b = train.loc[:, [best_feature_1b]]
features_test_1b = test.loc[:, [best_feature_1b]]

y_pred_1b, w_1b = train_1a(features_train_1b.to_numpy(), y_train.to_numpy(), features_test_1b.to_numpy())

In [87]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_personality_feature_model
mae_1b, w_1b = question_1a_MAE_check(y_test.to_numpy(), y_pred_1b, w_1b)

print("Weight:\n", np.round(w_1b, 3), "\nMAE:", np.round(mae_1b, 3))

Weight:
 [[-56546.304]] 
MAE: 291019.693


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} =  (-56546.304) * nueroticism $$

# Yêu cầu 1c: Xây dựng mô hình sử dụng duy nhất 1 đặc trưng `English`, `Logical`, `Quant`, tìm mô hình cho kết quả tốt nhất (1 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ đặc trưng

In [88]:
# Phần code cho yêu cầu 1c
# Tìm ra đặc trưng tốt nhất
# In ra các kết quả cross-validation như yêu cầu
features_1c = [
    'English',
    'Logical',
    'Quant',
    'Salary'
]

train_1c = train[features_1c]

mae_arr_1c, best_feature_1c = question_1b_handler(train_1c, features_1c, 5)
print("Best feature: ", best_feature_1c)

MAE:  [122012.093 120338.137 118133.757]
Best feature:  Quant


In [89]:
# Huấn luyện lại mô hình best_skill_feature_model với đặc trưng tốt nhất trên toàn bộ tập huấn luyện
features_train_1c = train.loc[:, [best_feature_1c]]
features_test_1c = test.loc[:, [best_feature_1c]]

y_pred_1c, w_1c = train_1a(features_train_1c.to_numpy(), y_train.to_numpy(), features_test_1c.to_numpy())

In [90]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình best_skill_feature_model
mae_1c, w_1c = question_1a_MAE_check(y_test.to_numpy(), y_pred_1c, w_1c)

print("Weight:\n", np.round(w_1c, 3), "\nMAE:", np.round(mae_1c, 3))

Weight:
 [[585.895]] 
MAE: 106819.578


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = 585.895 * Quant $$

# Yêu cầu 1d: Sinh viên tự xây dựng mô hình, tìm mô hình cho kết quả tốt nhất (3 điểm)

Lưu ý: khi sử dụng cross-validation, sinh viên cần xáo trộn dữ liệu 1 lần duy nhất và thực hiện trên toàn bộ $m$ mô hình mà sinh viên thiết kế

## Tìm mô hình

In [91]:
# Trình bày các phần tìm ra mô hình

In [92]:
def data_constructor(data, option=0):
    if(option == 0): # based on grades at university
        return data.loc[:, [
            "collegeGPA",
            "ComputerScience",
            "ComputerProgramming",
            "ElectronicsAndSemicon",
            "TelecomEngg",
            "Salary",
        ]]
    elif(option == 1): # based on all AMCAT scores and sqrt of numbers directly related to professional skills
        return pd.DataFrame([[np.sqrt(np.abs(d[i])) if i == 3 or i == 4 or i == 5 or i == 6 or i == 7 or i == 8 or i == 9 or i == 10
                               else d[i] for i in range(17)] for d in ((data.loc[:,[
            "English",
            "Logical",
            "Quant",
            "Domain", # sqrt
            "ComputerProgramming", # sqrt
            "ElectronicsAndSemicon", # sqrt
            "ComputerScience", # sqrt
            "MechanicalEngg", # sqrt
            "ElectricalEngg", # sqrt
            "TelecomEngg", # sqrt
            "CivilEngg", # sqrt
            "conscientiousness",
            "agreeableness",
            "extraversion",
            "nueroticism",
            "openess_to_experience",
            "Salary",
        ]
        ]).to_numpy())], columns = [
            "English",
            "Logical",
            "Quant",
            "Domain",
            "ComputerProgramming",
            "ElectronicsAndSemicon",
            "ComputerScience",
            "MechanicalEngg",
            "ElectricalEngg",
            "TelecomEngg",
            "CivilEngg",
            "conscientiousness",
            "agreeableness",
            "extraversion",
            "nueroticism",
            "openess_to_experience",
            "Salary",
        ])
    elif(option == 2): # based on all AMCAT scores and squares of numbers directly related to professional skills
        return pd.DataFrame([[d[i] ** 2 if i == 1 or i == 3 or i == 4 or i ==5 or i == 6 or i == 7 or i == 8 or i == 9 or i == 10 or i == 13
                               else d[i] for i in range(17)] for d in ((data.loc[:,[
            "English",
            "Logical", # ** 2
            "Quant",
            "Domain", # ** 2
            "ComputerProgramming", # ** 2
            "ElectronicsAndSemicon", # ** 2
            "ComputerScience", # ** 2
            "MechanicalEngg", # ** 2
            "ElectricalEngg", # ** 2
            "TelecomEngg", # ** 2
            "CivilEngg", # ** 2
            "conscientiousness",
            "agreeableness",
            "extraversion", # ** 2
            "nueroticism",
            "openess_to_experience",
            "Salary",
        ]
        ]).to_numpy())], columns = [
            "English",
            "Logical",
            "Quant",
            "Domain",
            "ComputerProgramming",
            "ElectronicsAndSemicon",
            "ComputerScience",
            "MechanicalEngg",
            "ElectricalEngg",
            "TelecomEngg",
            "CivilEngg",
            "conscientiousness",
            "agreeableness",
            "extraversion",
            "nueroticism",
            "openess_to_experience",
            "Salary",
        ])
    elif(option == 3): # based on AMCAT major scores and squares of scores of the first 4 parameters
        return pd.DataFrame([[d[i] ** 2 if i == 0 or i == 1 or i == 2 or i == 3
                               else d[i] for i in range(12)] for d in ((data.loc[:,[
            "English", # ** 2
            "Logical", # ** 2
            "Quant", # ** 2
            "Domain", # ** 2
            "ComputerProgramming", 
            "ElectronicsAndSemicon",
            "ComputerScience",
            "MechanicalEngg",
            "ElectricalEngg",
            "TelecomEngg",
            "CivilEngg",
            "Salary",
        ]
        ]).to_numpy())], columns = [
            "English",
            "Logical",
            "Quant",
            "Domain",
            "ComputerProgramming",
            "ElectronicsAndSemicon",
            "ComputerScience",
            "MechanicalEngg",
            "ElectricalEngg",
            "TelecomEngg",
            "CivilEngg",
            "Salary",
        ])
    elif(option == 4): # based on scores of the previous exams
        return data.loc[:, [
            "10percentage",
            "12percentage",
            "collegeGPA",
            "Salary",
        ]]
    elif(option == 5): # based on only collegeGPA
        return data.loc[:, [
            "collegeGPA",
            "Salary",
        ]]
    else:
        print('Error option')
        return data
    
def question_1d_handler(train, k_fold = 5, number_of_models = 5):
    mae_arr = np.zeros((number_of_models, 1))
    shuffle_data =  KFold(n_splits=k_fold, shuffle=True)

    for train_split, test_split in shuffle_data.split(train):
        # print("Train split", train_split, "Test split", test_split)
        for i in range(number_of_models):

            model_data = data_constructor(train, i)
            # print("Loop", i, model_data)

            features_train = model_data.iloc[train_split].iloc[:, :-1]
            labels_train = model_data.iloc[train_split].loc[:,['Salary']]

            features_test = model_data.iloc[test_split].iloc[:, :-1]
            labels_test = model_data.iloc[test_split].loc[:,['Salary']]
                            
            y_pred, w = train_1a(features_train.to_numpy(), labels_train.to_numpy(), features_test.to_numpy())
            mae, w = question_1a_MAE_check(labels_test.to_numpy(), y_pred, w)

            mae_arr[i] += mae
            i += 1
            
    mae_arr /= k_fold

    print('MAE:\n', np.round(mae_arr, 3))
    return mae_arr, np.argmin(mae_arr)

## Thử nghiệm, so sánh các mô hình

In [93]:
# Phần code cho yêu cầu 1d
# Tìm ra mô hình tốt nhất (tự thiết kế bởi sinh viên)
# In ra các kết quả cross-validation như yêu cầu

m = 6 # models

mae_arr_1d, best_1d = question_1d_handler(train, k_fold = 5, number_of_models = m)

print("My best model: number", best_1d + 1)

MAE:
 [[119226.88 ]
 [113076.154]
 [112830.892]
 [115745.199]
 [118250.137]
 [121468.404]]
My best model: number 3


In [94]:
# Huấn luyện lại mô hình my_best_model trên toàn bộ tập huấn luyện
features_train_1d = data_constructor(train, best_1d).iloc[:, :-1]
features_test_1d = data_constructor(test, best_1d).iloc[:, :-1]

# print(np.round(features_train_1d, 3))

y_pred_1d, w_1d = train_1a(features_train_1d.to_numpy(), y_train.to_numpy(), features_test_1d.to_numpy())

In [95]:
# Gọi hàm MAE (tự cài đặt hoặc từ thư viện) trên tập kiểm tra với mô hình my_best_model
mae_1d, w_1d = question_1a_MAE_check(y_test.to_numpy(), y_pred_1d, w_1d)

print("Weight:\n", w_1d, "\nMAE:", np.round(mae_1d, 3))

Weight:
 [[ 2.29513334e+02]
 [ 1.53858283e-01]
 [ 2.00944668e+02]
 [ 3.55669852e+04]
 [ 2.49028680e-01]
 [ 4.70730811e-03]
 [-3.62752712e-01]
 [ 2.34047135e-01]
 [-1.74573569e-01]
 [-7.78367974e-02]
 [ 5.60902389e-01]
 [-1.88116431e+04]
 [ 1.72035162e+04]
 [-1.94204514e+03]
 [-8.76388890e+03]
 [-6.98238289e+03]] 
MAE: 104739.987


Công thức hồi quy (phần trọng số làm tròn đến 3 chữ số thập phân, ví dụ 0.012345 $\to$ 0.012)

$$\text{Salary} = 229.513 * English + 153.858 * \text{Logical} ^ 2 + 200.945 * Quant + 355.670 * \text{Domain} ^ 2 \\
        + 249.029 * \text{ComputerProgramming} ^ 2 + 470.731 *  \text{ElectronicsAndSemicon} ^ 2 \\
        + (-362.753) * \text{ComputerScience} ^ 2 + 234.047 * \text{MechanicalEngg} ^ 2 + (-174.574) * \text{ElectricalEngg} ^ 2 \\
        + (-778.368) * \text{TelecomEngg} ^ 2 + 560.902 * \text{CivilEngg} ^ 2 + (-188.116) * conscientiousness \\
        + 172.035 * agreeableness + (-194.205) * \text{extraversion} ^ 2 + (-876.389) * nueroticism \\
        + (-698.238) * openess\_to\_experience
$$